QuestionAnswering/QA_SEQ2SEQ/data_generation/locate_then_ask/singapore/ask.py

```diff
@@ -0,0 +1,191 @@
+        OPltPlotAttrKey.GROSS_FLOOR_AREA: "om:squareMetre",
+        OPltPlotAttrKey.PLOT_AREA: "om:squareMetre",
+    }
+
+    def __init__(self):
+        self.graph2sparql = Graph2Sparql()
+
+    def _find_unsampled_keys(self, query_graph: QueryGraph):
+        sampled_attr_keys = tuple(
+            key for _, _, key in query_graph.edges(data="key") if key is not None
+        )
+        return tuple(
+            x
+            for x in OPltPlotAttrKey
+            if x not in sampled_attr_keys
+            and x is not OPltPlotAttrKey.IS_AWAITING_DETAILED_GPR_EVAL
+        )
+
+    def _find_unsampled_numerical_keys(self, query_graph: QueryGraph):
+        sampled_attr_keys = tuple(
+            key for _, key in query_graph.nodes(data="key") if key is not None
+        )
+        return [k for k in self.NUMERICAL_KEYS if k not in sampled_attr_keys]
+
+    def _ask_attr(self, query_graph: QueryGraph, key: OPltPlotAttrKey):
+        query_graph.add_question_node(key.value)
+        query_graph.add_triple("Plot", PLOT_ATTR_2_PRED[key], key.value)
+        verbn = random.choice(PLOT_ATTR_LABELS[key])
+
+        if key is OPltPlotAttrKey.GROSS_PLOT_RATIO:
+            # ask also is_awaiting_detailed_gpr_evaluation
+            key2 = OPltPlotAttrKey.IS_AWAITING_DETAILED_GPR_EVAL
+            query_graph.add_question_node(key2.value)
+            query_graph.add_triple("Plot", PLOT_ATTR_2_PRED[key2], key2.value)
+
+        return verbn
+
+    def ask_count(self, query_graph: QueryGraph, verbalization: str):
+        query_graph.add_question_node("Plot", agg=AggOp.COUNT)
+
+        query_sparql = self.graph2sparql.convert(query_graph)
+        template = random.choice(
+            [
+                "How many {located} are there",
+                "For {located}, how many of them are there",
+            ]
+        )
+        verbalization = template.format(located=verbalization)
+
+        return query_sparql, verbalization
+
+    def ask_attrs(self, query_graph: QueryGraph, verbalization: str, attr_num: int = 1):
+        unsampled_attr_keys = self._find_unsampled_keys(query_graph)
+
+        verbns = []
+        for key in random.sample(
+            unsampled_attr_keys, k=min(attr_num, len(unsampled_attr_keys))
+        ):
+            verbn = self._ask_attr(query_graph, key)
+            verbns.append(verbn)
+
+        query_sparql = self.graph2sparql.convert(query_graph)
+        template = random.choice(
+            [
+                "What is the {attrs} of the {located}",
+                "For the {located}, what is its {attrs}",
+            ]
+        )
+        verbalization = template.format(
+            attrs=" and ".join(verbns), located=verbalization
+        )
+
+        return query_sparql, verbalization
+
+    def ask_agg(self, query_graph: QueryGraph, verbalization: str, attr_num: int = 1):
+        candidates = self._find_unsampled_numerical_keys(query_graph)
+        assert len(candidates) > 0
```

Comment:
See above comment.

Reply:
I've replaced the `assert` statement with `raise ValueError(...)`.